pkg/admission/reservednames/admission_test.go

```diff
@@ -0,0 +1,105 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reservednames
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+)
+
+func createAttr(name string, obj runtime.Object, kind, resource string) admission.Attributes {
+	return admission.NewAttributesRecord(
+		obj,
+		obj,
+		tenancyv1alpha1.Kind(kind).WithVersion("v1alpha1"),
+		"",
+		name,
+		tenancyv1alpha1.Resource(resource).WithVersion("v1alpha1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func TestAdmission(t *testing.T) {
+	cases := map[string]struct {
+		attr admission.Attributes
+		want string
+	}{
+		"ForbiddenRootCW": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "ClusterWorkspace", "clusterworkspaces"),
+			want: fmt.Sprintf("cannot use reserved name \"%s\" as \"metadata.name\"", "root"),
+		},
+		"ForbiddenSystemCW": {
+			attr: createAttr("system", &tenancyv1alpha1.ClusterWorkspace{}, "ClusterWorkspace", "clusterworkspaces"),
+			want: fmt.Sprintf("cannot use reserved name \"%s\" as \"metadata.name\"", "system"),
+		},
+		"ValidCW": {
+			attr: createAttr("cool-cw", &tenancyv1alpha1.ClusterWorkspace{}, "ClusterWorkspace", "clusterworkspaces"),
+		},
+		"ForbiddenRootCWT": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspaceType{}, "ClusterWorkspaceType", "clusterworkspacetypes"),
+			want: fmt.Sprintf("cannot use reserved name \"%s\" as \"metadata.name\"", "root"),
+		},
+		"ForbiddenSystemCWT": {
+			attr: createAttr("system", &tenancyv1alpha1.ClusterWorkspaceType{}, "ClusterWorkspaceType", "clusterworkspacetypes"),
+			want: fmt.Sprintf("cannot use reserved name \"%s\" as \"metadata.name\"", "system"),
+		},
+		"ValidCWT": {
+			attr: createAttr("cool-cwt", &tenancyv1alpha1.ClusterWorkspaceType{}, "ClusterWorkspaceType", "clusterworkspacetypes"),
+		},
+		"NotApplicableResource": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "ClusterWorkspace", "notacworcwt"),
+		},
+		"NotApplicableKind": {
+			attr: createAttr("root", &tenancyv1alpha1.ClusterWorkspace{}, "NotaCWorCWT", "clusterworkspaces"),
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			plugin := &reservedNames{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+				applicableKinds: map[schema.GroupKind]struct{}{
```

Comment:
Maybe we should make a constructor for the `reservedNames` instance and call that here?